File: gocode_gui/lib/core/api/account.dart
/// The gocoder.org account behind the server's sign-in (GET /api/account).
/// The server calls the site with its stored API key, so the app never holds
/// a gocoder.org credential itself.
class AccountInfo {
  const AccountInfo({
    required this.signedIn,
    required this.site,
    this.userID = '',
    this.email = '',
    this.displayName = '',
    this.keyPrefix = '',
    this.memberSince,
    this.expired = false,
    this.offline = false,
  });

  factory AccountInfo.fromJson(Map<String, dynamic> json) => AccountInfo(
    signedIn: json['signedIn'] as bool? ?? false,
    site: json['site'] as String? ?? '',
    userID: json['userId'] as String? ?? '',
    email: json['email'] as String? ?? '',
    displayName: json['displayName'] as String? ?? '',
    keyPrefix: json['keyPrefix'] as String? ?? '',
    memberSince: DateTime.tryParse(json['memberSince'] as String? ?? '')
        ?.toLocal(),
    expired: json['expired'] as bool? ?? false,
    offline: json['offline'] as bool? ?? false,
  );

  final bool signedIn;

  /// The gocoder.org instance, e.g. `https://gocoder.org`.
  final String site;
  final String userID;
  final String email;
  final String displayName;

  /// This machine's API key prefix; the key itself stays on the server.
  final String keyPrefix;
  final DateTime? memberSince;

  /// The site rejects the stored key (revoked, or the account was deleted).
  /// Signing in again fixes it.
  final bool expired;

  /// The site couldn't be reached; the details are the stored ones.
  final bool offline;

  /// The name to show: the display name, else the email's local part.
  String get name =>
      displayName.isNotEmpty ? displayName : email.split('@').first;

  /// Up to two initials for the avatar.
  String get initials {
    final words = name
        .trim()
        .split(RegExp(r'\s+'))
        .where((w) => w.isNotEmpty)
        .toList();
    if (words.isEmpty) return '?';
    String first(String word) =>
        String.fromCharCode(word.runes.first).toUpperCase();
    return words.length == 1
        ? first(words.first)
        : first(words.first) + first(words.last);
  }

  /// A page on the website: `page('/settings')` → `<site>/#/settings`.
  Uri page(String path) {
    final base = site.isEmpty
        ? 'https://gocoder.org'
        : site.replaceAll(RegExp(r'/+$'), '');
    return Uri.parse('$base/#$path');
  }
}

int _int(Object? value) => (value as num?)?.toInt() ?? 0;

/// Totals over one usage window.
class UsageTotals {
  const UsageTotals({
    this.requests = 0,
    this.tokens = 0,
    this.promptTokens = 0,
    this.cachedRead = 0,
    this.errors = 0,
    this.avgLatencyMs = 0,
  });

  factory UsageTotals.fromJson(Map<String, dynamic> json) => UsageTotals(
    requests: _int(json['requests']),
    tokens: _int(json['tokens']),
    promptTokens: _int(json['promptTokens']),
    cachedRead: _int(json['cachedRead']),
    errors: _int(json['errors']),
    avgLatencyMs: (json['avgLatencyMs'] as num?)?.toDouble() ?? 0,
  );

  final int requests;
  final int tokens;
  final int promptTokens;
  final int cachedRead;
  final int errors;
  final double avgLatencyMs;

  /// Share of prompt tokens served from cache; null when none were reported.
  double? get cacheHitRate =>
      promptTokens > 0 ? cachedRead / promptTokens : null;
}

/// One provider+model's share of a window.
class UsageModel {
  const UsageModel({
    required this.provider,
    required this.model,
    required this.requests,
    required this.tokens,
  });

  factory UsageModel.fromJson(Map<String, dynamic> json) => UsageModel(
    provider: json['provider'] as String? ?? '',
    model: json['model'] as String? ?? '',
    requests: _int(json['requests']),
    tokens: _int(json['tokens']),
  );

  final String provider;
  final String model;
  final int requests;
  final int tokens;
}

/// One zero-filled UTC day.
class UsageDay {
  const UsageDay({
    required this.date,
    required this.requests,
    required this.tokens,
  });

  factory UsageDay.fromJson(Map<String, dynamic> json) => UsageDay(
    date: DateTime.tryParse(json['date'] as String? ?? '') ?? DateTime(1970),
    requests: _int(json['requests']),
    tokens: _int(json['tokens']),
  );

  final DateTime date;
  final int requests;
  final int tokens;
}

/// The account's usage over a window of days (GET /api/account/usage).
class UsageSummary {
  const UsageSummary({
    required this.totals,
    required this.byModel,
    required this.daily,
  });

  factory UsageSummary.fromJson(Map<String, dynamic> json) => UsageSummary(
    totals: UsageTotals.fromJson(
      json['totals'] as Map<String, dynamic>? ?? const {},
    ),
    byModel: (json['byModel'] as List<dynamic>? ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(UsageModel.fromJson)
        .toList(),
    daily: (json['daily'] as List<dynamic>? ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(UsageDay.fromJson)
        .toList(),
  );

  final UsageTotals totals;
  final List<UsageModel> byModel;
  final List<UsageDay> daily;
}

/// The account's invite link and how many people joined through it.
class InviteInfo {
  const InviteInfo({
    required this.code,
    required this.url,
    required this.invited,
  });

  factory InviteInfo.fromJson(Map<String, dynamic> json) => InviteInfo(
    code: json['code'] as String? ?? '',
    url: json['url'] as String? ?? '',
    invited: _int(json['invited']),
  );

  final String code;
  final String url;
  final int invited;
}

File: internal/gocoder/account.go
package gocoder

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"strconv"
	"time"
)

// account.go holds what a signed-in gocode shows about its account: profile,
// usage and invite link, plus the sign-in bookkeeping the server's
// /api/account routes share with onboarding. The calls take the stored API
// key as bearer; the website accepts keys on these routes.

// KeyName names the API key a sign-in on this machine creates.
func KeyName() string {
	name := "gocode"
	if host, err := os.Hostname(); err == nil && host != "" {
		name += " on " + host
	}
	return name
}

// StoreSession trades a session's short-lived token for a named API key and
// saves the account (gocoder.json) the rest of gocode reads.
func StoreSession(ctx context.Context, client *Client, session *Session, keyName string) (*Account, error) {
	key, err := client.CreateKey(ctx, session.Token, keyName)
	if err != nil {
		return nil, err
	}
	account := &Account{
		URL:         client.BaseURL,
		UserID:      session.User.ID,
		Email:       session.User.Email,
		DisplayName: session.User.DisplayName,
		KeyID:       key.ID,
		KeyPrefix:   key.Prefix,
		Key:         key.Secret,
		CreatedAt:   time.Now(),
	}
	if err := SaveAccount(account); err != nil {
		return nil, fmt.Errorf("saving account: %w", err)
	}
	return account, nil
}

// Me returns the account's current profile.
func (c *Client) Me(ctx context.Context, bearer string) (*User, error) {
	var user User
	if err := c.do(ctx, http.MethodGet, "/api/auth/me", bearer, nil, &user); err != nil {
		return nil, err
	}
	return &user, nil
}

// CheckKey reports whether the site still accepts bearer as an API key: nil
// when it does, an *APIError with status 401 when the key is revoked. It
// asks the settings-sync revision route, which takes keys wherever the site
// is deployed — unlike /api/auth/me, whose 401 can't tell a revoked key from
// a deployment that wants a session token there.
func (c *Client) CheckKey(ctx context.Context, bearer string) error {
	_, err := c.SettingsRevision(ctx, bearer, "")
	return err
}

// UpdateDisplayName renames the account. Password changes are not offered:
// the website refuses them with an API key and wants a signed-in session.
func (c *Client) UpdateDisplayName(ctx context.Context, bearer, displayName string) (*User, error) {
	var user User
	body := map[string]string{"displayName": displayName}
	if err := c.do(ctx, http.MethodPatch, "/api/auth/me", bearer, body, &user); err != nil {
		return nil, err
	}
	return &user, nil
}

// UsageTotals aggregates one usage window (the website's /api/stats shape).
type UsageTotals struct {
	Requests     int     `json:"requests"`
	Tokens       int     `json:"tokens"`
	PromptTokens int     `json:"promptTokens"`
	CachedRead   int     `json:"cachedRead"`
	CachedWrite  int     `json:"cachedWrite"`
	Errors       int     `json:"errors"`
	AvgLatencyMs float64 `json:"avgLatencyMs"`
	ActiveKeys   int     `json:"activeKeys"`
}

// UsageProvider is one provider's share of a window.
type UsageProvider struct {
	Provider string `json:"provider"`
	Requests int    `json:"requests"`
	Tokens   int    `json:"tokens"`
}

// UsageModel is one provider+model's share of a window.
type UsageModel struct {
	Provider string `json:"provider"`
	Model    string `json:"model"`
	Requests int    `json:"requests"`
	Tokens   int    `json:"tokens"`
}

// UsageDay is one zero-filled UTC day.
type UsageDay struct {
	Date     string `json:"date"`
	Requests int    `json:"requests"`
	Tokens   int    `json:"tokens"`
}

// UsageSummary is the account's usage over a window of days.
type UsageSummary struct {
	Totals     UsageTotals     `json:"totals"`
	ByProvider []UsageProvider `json:"byProvider"`
	ByModel    []UsageModel    `json:"byModel"`
	Daily      []UsageDay      `json:"daily"`
}

// UsageSummary aggregates the account's last days days (the site allows 1-90).
func (c *Client) UsageSummary(ctx context.Context, bearer string, days int) (*UsageSummary, error) {
	var summary UsageSummary
	path := "/api/stats/summary?days=" + strconv.Itoa(days)
	if err := c.do(ctx, http.MethodGet, path, bearer, nil, &summary); err != nil {
		return nil, err
	}
	return &summary, nil
}

// Invite is the account's shareable invite: its code, the sign-up link that
// carries it, and how many people have joined through it.
type Invite struct {
	Code    string `json:"code"`
	URL     string `json:"url"`
	Invited int64  `json:"invited"`
}

// Invite returns the account's invite link and referral count.
func (c *Client) Invite(ctx context.Context, bearer string) (*Invite, error) {
	var invite Invite
	if err := c.do(ctx, http.MethodGet, "/api/invites", bearer, nil, &invite); err != nil {
		return nil, err
	}
	return &invite, nil
}
